app/models/miq_server/worker_management/heartbeat.rb

```diff
@@ -48,15 +48,10 @@ def message_for_worker(wid, message, *args)
 
   def post_message_for_workers(class_name = nil, resync_needed = false, sync_message = nil)
     processed_worker_ids = []
-    miq_workers.each do |w|
+
+    miq_workers.where(:status => MiqWorker::STATUSES_CURRENT_OR_STARTING).each do |w|
```

Comment:
adding the `.where` here forces activerecord to query via the association instead of using the cached and possibly stale `.miq_workers`... I copied this from @carbonin fix in https://github.com/ManageIQ/manageiq/commit/e6cbfa8b4625147741919c723eacac540f056099 as it was cleaner than tracking "possible" changes in the association and doing a conditional reload

Reply:
But this is now an unconditional reload, right?

I guess I would rather just have a `miq_workers.reload` in the main loop to make it look like it was intentional. This feels like a side-effect of how we're querying. Say, for example, we decided in the future that we actually want to query for all workers here and remove the `where`, we would end up with this bug again, right? I don't think I would think that we needed the where to reload the association as well as filtering the workers we want to iterate through.

Reply:
yes, it's an unconditional reload here now... hmm, interesting, I thought the `where` made it more explicit about wanting to query it and not possibly use a cached association..  I can change it back to track the result of validate_worker and conditionally reload the association if desired.

Reply:
Ok, I put a local `reload_workers` variable to track the result of `validate_worker` and any were restarted, we'll reload the miq_workers association.